src/frontend/src/flows/addDevice/welcomeView/registerTentativeDevice.ts

```diff
@@ -1,174 +1,168 @@
+    };
+  const addResponse = await addTentativeDevice({
+    userNumber,
+    connection,
+    device,
+  });
 
-  const continueButton = document.getElementById(
-    "registerTentativeDeviceContinue"
-  ) as HTMLButtonElement;
-  const aliasInput = document.getElementById(
-    "tentativeDeviceAlias"
-  ) as HTMLInputElement;
+  // If everything went well we can now ask the user to authenticate on an existing device
+  // and enter a verification code
+  await showVerificationCode(
+    userNumber,
+    connection,
+    device.alias,
+    addResponse.added_tentatively,
+    device.credential_id[0]
+  );
+};
 
-  aliasInput.onkeypress = (e) => {
-    // submit if user hits enter
-    if (e.key === "Enter") {
-      e.preventDefault();
-      continueButton.click();
+/** Create new WebAuthn credentials */
+const createDevice = async ({
+  userNumber,
+  connection,
+}: {
+  userNumber: bigint;
+  connection: Connection;
+}): Promise<WebAuthnIdentity | Error> => {
+  const existingAuthenticators = await connection.lookupAuthenticators(
+    userNumber
+  );
+  let newDevice: WebAuthnIdentity;
+  try {
+    newDevice = await WebAuthnIdentity.create({
+      publicKey: creationOptions(existingAuthenticators),
+    });
+  } catch (error: unknown) {
+    if (error instanceof Error) {
+      return error;
+    } else {
+      return new Error(unknownToString(error, "unknown error"));
     }
-  };
+  }
+  return newDevice;
+};
 
-  continueButton.onclick = async () => {
-    if (aliasInput.value === "") {
-      toggleErrorMessage("tentativeDeviceAlias", "invalidAliasMessage", true);
-      return;
-    }
+type AddDeviceSuccess = Extract<
+  AddTentativeDeviceResponse,
+  { added_tentatively: unknown }
+>;
 
-    let newDevice: WebAuthnIdentity;
-    try {
-      newDevice = await WebAuthnIdentity.create({
-        publicKey: creationOptions(existingAuthenticators),
-      });
-    } catch (error: unknown) {
+/** Add the device tentatively to the canister */
+export const addTentativeDevice = async ({
+  userNumber,
+  connection,
+  device,
+}: {
+  userNumber: bigint;
+  connection: Connection;
+  device: Omit<DeviceData, "origin">;
+}): Promise<AddDeviceSuccess> => {
+  // Try to add the device tentatively, retrying if necessary
+  for (;;) {
+    const result = await withLoader(() =>
+      connection.addTentativeDevice(userNumber, device)
+    );
+
+    if ("another_device_tentatively_added" in result) {
+      // User already added a device so we show an error and abort
       await displayError({
-        title: "Error adding new device",
-        message: "Unable to register new WebAuthn Device.",
-        detail: error instanceof Error ? error.message : JSON.stringify(error),
+        title: "Tentative Device Already Exists",
+        message:
+          'The "add device" process was already started for another device. If you want to add this device instead, log in using an existing device and restart the "add device" process.',
```

Comment:
nth: there is the word "device" six times in this sentence, don't know if can be improved?

as you know my English is so good, I'm the best to make these kinds of comment 😉

Reply:
Hahaha, not my call unfortunately. Will let you take it up with @marydwyer

Reply:
Haha true, though this was just moved around, I didn't change the sentence. I'll leave it as is

CC @marydwyer for suggestions